### ISSUE TITLE
Implementation of ndimage.filters.gaussian_filter

### ISSUE THREAD

**Issue by aymgal** [jax-ml/jax#7284]
Hi, can we expect to have the fast [gaussian_filter](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.gaussian_filter.html) method from scipy implemented in JAX? Or are their fundamentals limitations that prevent its implementation?

**Comment by shoyer** [comment]
It looks like scipy implements a N-D Gaussian filter via N separate 1D filters over each dimension, which are in turn implemented via 1D convolutions.

JAX has all of these building blocks already, so a Gaussian filter should be straightforward to implement with similar code to what SciPy uses.

One limitation is that it won't be very efficient for large radius blurs, because JAX doesn't (yet) implement FFT based convolutions. See https://github.com/google/jax/issues/5227 and https://github.com/google/jax/pull/6343 for discussion and progress on that.

**Comment by hawkinsp** [comment]
Another option would be to contribute these functions to a library like `dm_pix`, which was recently opensourced: https://github.com/deepmind/dm_pix

Happy either way.

**Comment by aymgal** [comment]
Thank you both for the suggestions. The option proposed by @shoyer looks like the way to go, it seems more efficient compared to what's currently in `dm_pix`. Unfortunately I have currently no time to get my hands dirty and implement this (in a proper way) to `jax` anytime soon... Do we keep this issue open?

**Comment by carcablop** [comment]
Hello.I am an Outreachy applicant and I would like to contribute to this project. Could you work on this bug?, Is it still available to work? Or could someone help me with a problem that is a good start to contribute?. Thanks.

**Comment by jakevdp** [comment]
Hi @carcablop - I was the one who originally proposed the outreachy project, but I withdrew the proposal once I realized that the outreachy application model relies on having dozens of entry-level issues to point people to. We simply don't have many such issues, so I realized JAX is not a great fit for outreachy.

I would consider the current issue an appropriate entry-point for someone with experience implementing operations in a vector-based language like NumPy. If it's something you'd like to tackle, I'd suggest looking into how existing `ndimage` functions are implemented (see https://github.com/google/jax/blob/main/jax/_src/scipy/ndimage.py) and tested (see https://github.com/google/jax/blob/main/tests/scipy_ndimage_test.py), and model a `gaussian_filter` implementation based on that.

Hope that helps!

**Comment by kimoyerr** [comment]
Hey @carcablop or @jakevdp. Is this issue still active? If so, I would like to take a swing at it. Thank you

**Comment by renecotyfanboy** [comment]
Hi everyone, I would be super interested and very glad to implement the gaussian blur in JAX since I already did this specific task for research. 

However, I see there is not yet any strong decision about the way it should be done. There are basically 2 ways to make this, the FFT approach and the direct convolution. Since I am not very into NN, I have pretty much no idea about how the LAX implemented convolution works and performs when compared to FFT in this very specific task. (I will probably end benchmarking both and see which one performs best). 

I may favour the FFT approach since it might be differentiable w.r.t the kernel size without significant loss of performance, while convolution usually requires having the kernel size depending on sigma (and I think that JAX doesn't like array size that depends on input parameters).

Any advice is very welcome!

**Comment by jakevdp** [comment]
We have decided that `scipy.ndimage` is out-of-scope for the JAX core project:  https://jax.readthedocs.io/en/latest/jep/18137-numpy-scipy-scope.html#scipy-ndimage 

If anyone is interested in implementing these kinds of routines, we'd suggest developing them in an independent project.

**Comment by Jhko725** [comment]
I'm late to the party, but just in case anybody needs this function and comes across this thread, here is a simple implementation corresponding to the case of `order=1`

```python
import jax.numpy as jnp
from jax.scipy.signal import correlate
from jaxtyping import Array, Float


def gaussian_kernel_1d(sigma: float, radius: int) -> Float[Array, " 2*{radius}"]:
    """JAX version of the gaussian kernel to be used in the gaussian_filter1d function.

    Note that the implementation parallels that of scipy.ndimage at
    https://github.com/scipy/scipy/blob/v1.16.2/scipy/ndimage/_filters.py#L656."""
    x = jnp.arange(-radius, radius + 1)
    sigma2 = sigma * sigma
    phi_x = jnp.exp(-0.5 * x**2 / sigma2)
    return phi_x / jnp.sum(phi_x)


def gaussian_filter1d(
    x: Float[Array, " N"], sigma: float, truncate: float = 4.0
) -> Float[Array, " N"]:
    """JAX implementation of scipy.ndimage.gaussian_filter1d for order=1.

    Since jax.scipy.signal.correlate does not provide reflect padding, this is
    manually done via jax.numpy.pad."""
    radius = int(truncate * sigma + 0.5)
    weights = gaussian_kernel_1d(sigma, radius)[::-1]
    x_padded = jnp.pad(x, (radius, radius), mode="reflect")

    return correlate(x_padded, weights, mode="valid", method="fft")
```
The axis argument is not needed here, since you can achieve the same functionality via `jax.vmap`.

To test the function, you can do
```python
import jax
import numpy as np
import scipy.ndimage as scimage
from dynamical_systems.metrics.statistical import gaussian_filter1d
from scipy.signal import square

jax.config.update("jax_enable_x64", True)

def test_gaussian_filter1d():
    t = np.linspace(0, 1, 500, endpoint=False)
    waveform = square(2 * np.pi * 2 * t)
    waveform_scipy = scimage.gaussian_filter1d(waveform, 20)
    waveform_jax = gaussian_filter1d(waveform, 20)
    assert np.allclose(waveform_scipy, waveform_jax) # Should pass as long as both numpy and jax ran on the same precision
```

Extending this function to higher orders shouldn't be much of a hassle as one can directly port the original scipy implementation at https://github.com/scipy/scipy/blob/v1.16.2/scipy/ndimage/_filters.py#L656
